core/predicate_check.go

```diff
@@ -0,0 +1,49 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package core
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/subnet-evm/core/types"
+	"github.com/ava-labs/subnet-evm/params"
+	"github.com/ava-labs/subnet-evm/precompile/contract"
+	"github.com/ava-labs/subnet-evm/precompile/modules"
+	"github.com/ava-labs/subnet-evm/utils"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// CheckPredicates checks that all precompile predicates are satisfied within the current [predicateContext] for [tx]
+func CheckPredicates(rules params.Rules, predicateContext *contract.PredicateContext, tx *types.Transaction) error {
+	precompileConfigs := rules.ActivePrecompiles
+	// Track addresses that we've performed a predicate check for
+	precompileAddressChecks := make(map[common.Address]struct{})
+	for _, accessTuple := range tx.AccessList() {
+		address := accessTuple.Address
+		_, isPrecompile := precompileConfigs[address]
+		if !isPrecompile {
+			continue
+		}
+
+		module, ok := modules.GetPrecompileModuleByAddress(address)
+		if !ok {
+			return fmt.Errorf("predicate accessed precompile config under address %s with no registered module for tx %s", address, tx.Hash())
+		}
+		predicater, ok := module.Contract.(contract.Predicater)
+		if !ok {
+			continue
+		}
+		// Return an error if we've already checked a predicate for this address
+		if _, ok := precompileAddressChecks[address]; ok {
```

Comment:
just want to confirm that for a transaction, each precompile address' predicate should only appear in one access tuple. As in, no multiple access tuples of one transaction can specify the same precompile address.

Reply:
Yup that's correct. This is an intentional change from the previous code based on a conversation with @darioush to avoid a potentially weird situation where we verify all of the predicates and only expose the data from the last one.

It seemed pretty jank to overwrite the predicate storage slots with whatever the last item in the access list was, so this is intended to remove that edge case.

Reply:
that makes sense, I like the change from overwriting last item.